services/github/test/handler_test.js

```diff
@@ -407,6 +407,83 @@ helper.secrets.mockSuite(testing.suiteName(), [], function(mock, skipping) {
       assert.equal(build.state, 'pending');
     });
 
+    test('valid pull_request (user is not a collaborator) does not create tasks', async function() {
+      github.inst(5828).setTaskclusterYml({
+        owner: 'TaskclusterRobot',
+        repo: 'hooks-testing',
+        ref: '03e9577bc1ec60f2ff0929d5f1554de36b8f48cf', // HEAD
+        content: require('./data/yml/valid-yaml.json'),
+      });
+      github.inst(5828).setTaskclusterYml({
+        owner: 'TaskclusterRobot',
+        repo: 'hooks-testing',
+        ref: 'development', // default branch
+        content: require('./data/yml/valid-yaml.json'),
+      });
+
+      await simulateJobMessage({ user: 'goodBuddy', eventType: 'pull_request.opened' });
+
+      assert(handlers.createTasks.notCalled);
+    });
+
+    test('valid pull_request (user is not a collaborator, policy is public) creates a taskGroup', async function() {
+      let tcyaml = { ...require('./data/yml/valid-yaml-v1.json') };
+      tcyaml['policy'] = { 'pullRequests': 'public' };
+
+      github.inst(5828).setTaskclusterYml({
+        owner: 'TaskclusterRobot',
+        repo: 'hooks-testing',
+        ref: '03e9577bc1ec60f2ff0929d5f1554de36b8f48cf', // HEAD
+        content: tcyaml,
+      });
+      github.inst(5828).setTaskclusterYml({
+        owner: 'TaskclusterRobot',
+        repo: 'hooks-testing',
+        ref: 'development', // default branch
+        content: tcyaml,
+      });
+
+      await simulateJobMessage({ user: 'goodBuddy', eventType: 'pull_request.opened' });
+
+      assert(handlers.createTasks.calledWith({ scopes: sinon.match.array, tasks: sinon.match.array }));
+      let args = handlers.createTasks.firstCall.args[0];
+      let taskGroupId = args.tasks[0].task.taskGroupId;
+      let [build] = await helper.db.fns.get_github_build(taskGroupId);
+      assert.equal(build.organization, 'TaskclusterRobot');
+      assert.equal(build.repository, 'hooks-testing');
+      assert.equal(build.sha, '03e9577bc1ec60f2ff0929d5f1554de36b8f48cf');
+      assert.equal(build.state, 'pending');
+    });
+
+    test('valid pull_request (user is not a collaborator, policy is public_restricted) creates a taskGroup', async function() {
```

Comment:
It looks like both tests are identical with only change being `tcyaml['policy']` value?
And if I understand correctly, for the `public_restricted`, `createTasks()` would be called with different role, is it possible to add this check also? Or scopes do not go through the `createTasks` call directly?

Reply:
Yeah good call, I'll add assertions for the scopes!

And yes the tests are basically the same. I tend to prefer duplicating tests so that you can tell at a glance what's broken and what's not, but if you like I could try refactoring them to reduce duplication.

Reply:
Sounds good, turns out the `simulateJobMessage` helper wasn't setting the `payload.tasks_for` property, and once I add that some other tests start failing. Trying to dig into why and whether those failures are legit bugs or not.

Reply:
Tracked it down, turns out setting the `tasks_for` just unlocked a code path that needed an additional mocked property in the event body.